### PR TITLE
Add model management to RAGFlow HTTP API

### PR DESCRIPTION
**Pull request by as-ondewo**

### CODE REVIEW

File: api/utils/validation_utils.py [hunk 1]
```diff
@@ -723,3 +725,201 @@ def validate_id(cls, v: Any) -> str:
...
+    mineru_delete_output: Annotated[str | None, Field(default=None)]
+
+    @model_validator(mode="after")
+    def validate_model_addition_mode(self) -> "AddModelReq":
+        """
+        Validates the request based on addition mode:
+        - Factory-level: requires llm_factory and api_key (or special auth fields)
+        - Individual model: requires llm_factory, llm_name, model_type, and api_base (for local models) or api_key
+        """
+        from common.constants import LLMType
```
Comment (as-ondewo) [inline api/utils/validation_utils.py:797]: only use module level imports

File: api/utils/validation_utils.py [hunk 2]
```diff
@@ -723,3 +725,201 @@ def validate_id(cls, v: Any) -> str:
...
+
+        # Individual model addition mode
+        if self.llm_name is not None or self.model_type is not None:
+            if not self.llm_name:
+                raise PydanticCustomError("field_required", "llm_name is required when adding an individual model")
+            if not self.model_type:
+                raise PydanticCustomError("field_required", "model_type is required when adding an individual model")
+
+            # Validate model_type
+            valid_types = [LLMType.CHAT, LLMType.EMBEDDING, LLMType.RERANK, LLMType.IMAGE2TEXT, LLMType.SPEECH2TEXT, LLMType.TTS, LLMType.OCR]
```
Comment (as-ondewo) [inline api/utils/validation_utils.py:826]: List is unnecessary. To check if model type is valid use `self.model_type in LLMType`. To get list of `LLMType`s use `[t.value for t in LLMType]`

File: api/utils/validation_utils.py [hunk 3]
```diff
@@ -723,3 +725,201 @@ def validate_id(cls, v: Any) -> str:
...
+            elif self.llm_factory == "XunFei Spark":
+                if self.model_type == "chat":
+                    if not self.spark_api_password:
+                        raise PydanticCustomError("field_required", "spark_api_password is required for XunFei Spark chat models")
+                elif self.model_type == "tts":
+                    if not self.spark_app_id or not self.spark_api_secret or not self.spark_api_key:
+                        raise PydanticCustomError("field_required", "spark_app_id, spark_api_secret, and spark_api_key are required for XunFei Spark TTS models")
+
+        # Factory-level addition mode
+        else:
```
Comment (as-ondewo) [inline api/utils/validation_utils.py:865]: When adding all models of a factory only check factories for which this is actually supported. For all other's check that the API key is provided

File: api/utils/validation_utils.py [hunk 4]
```diff
@@ -723,3 +725,201 @@ def validate_id(cls, v: Any) -> str:
...
+                raise PydanticCustomError("field_required", "model_type is required when adding an individual model")
+
+            # Validate model_type
+            valid_types = [LLMType.CHAT, LLMType.EMBEDDING, LLMType.RERANK, LLMType.IMAGE2TEXT, LLMType.SPEECH2TEXT, LLMType.TTS, LLMType.OCR]
+            if self.model_type not in valid_types:
+                raise PydanticCustomError("invalid_value", f"model_type must be one of: {', '.join(valid_types)}")
+
+            # For local models, api_base is typically required
+            if is_local and not self.api_base and not self.api_key:
+                raise PydanticCustomError("field_required", "api_base is required for local/self-hosted models when api_key is not provided")
```
Comment (as-ondewo) [inline api/utils/validation_utils.py:832]: For local models the base URL should always be required even if an API key is provided

File: api/utils/validation_utils.py [hunk 5]
```diff
@@ -723,3 +725,201 @@ def validate_id(cls, v: Any) -> str:
...
+
+            # For individual model mode, validate factory-specific required parameters
```
Comment (as-ondewo) [inline api/utils/validation_utils.py:834]: use match-case construct

File: api/utils/validation_utils.py [hunk 6]
```diff
@@ -723,3 +725,201 @@ def validate_id(cls, v: Any) -> str:
...
+            if self.llm_factory == "VolcEngine":
+                if not self.ark_api_key or not self.endpoint_id:
+                    raise PydanticCustomError("field_required", "ark_api_key and endpoint_id are required for VolcEngine individual model addition")
```
Comment (as-ondewo) [inline api/utils/validation_utils.py:837]: * create helper function for these checks to reduce code duplication
* should also check that the fields don't just contain whitespace

File: api/utils/validation_utils.py [hunk 7]
```diff
@@ -723,3 +725,201 @@ def validate_id(cls, v: Any) -> str:
...
+                        self.google_project_id,
+                        self.api_version,
+                        self.provider_order,
+                        self.spark_app_id,
+                        self.spark_api_password,
+                    ]
+                )
+                if not self.api_key and not has_special_auth:
+                    raise PydanticCustomError("field_required", "api_key or appropriate authentication fields are required for factory-level addition")
+            # For local factories, empty api_key is allowed in factory-level mode (validation skipped above)
```
Comment (as-ondewo) [inline api/utils/validation_utils.py:923]: This comment is misleading. Local models aren't supported for factory-level adding

File: api/apps/sdk/users.py [hunk 8]
```diff
@@ -0,0 +1,850 @@
...
+                    "status": tenant_llm.get("status", "1"),
+                }
+            )
+
+    return get_result(data=res)
+
+
+@manager.route("/models", methods=["POST"])  # noqa: F821
+@token_required
+async def add_model(tenant_id: str) -> Response:
```
Comment (as-ondewo) [inline api/apps/sdk/users.py:249]: This function is full of duplicate validations that is already handled by the Pydantic request validator. Please rework it to remove the redundant code.

File: test/testcases/test_http_api/test_user_models_management/conftest.py [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+
+import pytest
+from pytest import FixtureRequest
+
+from common import remove_factory
+
+
+@pytest.fixture(scope="class")
+def cleanup_added_models(request: FixtureRequest, HttpApiAuth):
+    """Fixture to clean up models added during tests (for TestAddModel class)"""
```
Comment (as-ondewo) [inline test/testcases/test_http_api/test_user_models_management/conftest.py:26]: this is used for all tests not just `add_model`

File: test/testcases/test_http_api/test_user_models_management/conftest.py [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+    # Track factories that might be added during tests
+    factories_to_cleanup: List[str] = ["Builtin", "LocalAI", "Ollama", "Xinference", "LM-Studio", "GPUStack", "FastEmbed"]
```
Comment (as-ondewo) [inline test/testcases/test_http_api/test_user_models_management/conftest.py:28]: this is not an accurate list of the factories that are added in the tests

File: test/testcases/test_http_api/common.py [hunk 11]
```diff
@@ -247,3 +248,34 @@ def batch_add_sessions_with_chat_assistant(auth, chat_assistant_id, num):
         res = create_session_with_chat_assistant(auth, chat_assistant_id, {"name": f"session_with_chat_assistant_{i}"})
         session_ids.append(res["data"]["id"])
     return session_ids
+
+
+# USER MODELS MANAGEMENT
```
Comment (as-ondewo) [inline test/testcases/test_http_api/common.py:253]: The `headers` and `data` parameters of the functions here are used. Remove them.

File: api/apps/sdk/users.py [hunk 12]
```diff
@@ -0,0 +1,850 @@
...
+    """
```
Comment (as-ondewo) [inline api/apps/sdk/users.py:250]: The information in this docstring is wrong. Use the same docstring format as the other HTTP API endpoints. Additional details should be in the RAGFlow documentation.

File: test/testcases/test_http_api/test_user_models_management/test_add_model.py
Comment (as-ondewo) [inline test/testcases/test_http_api/test_user_models_management/test_add_model.py:1]: There are a lot of problems with the tests in this module, so I'll only focus on the parameter validation tests for now:
* when testing a string parameter that needs to be present test 4 case: not providing the parameter at all, passing `None`, an empty string, and a string with only whitespace
* make sure to construct your requests so that the reason they fail is related to what you are trying to test and not e.g. some unrelated missing parameter
* The tests that test factories with special parameters are disorganized and partially redundant. Organize them into one `TestSpecialParameterValidation` class containing the following parameterized tests:
  * for each factory that has special parameters and can be added in "factory-mode" test the following when adding in factory mode (no `llm_name` or `model_type`):
    * providing all required parameters passes the validation (but request is still expected to fail when testing if model can be accessed)
    * for each of the special parameter test not providing it
    * for each special parameter test that providing it which a factory where it is not expected fails
  * for all factories that have special parameters add the same tests as described above also for adding individual models
  * for each factory without special parameters test that validation passes for a minimal valid config with any special parameter when adding in both factory-mode (for factories that support it) and individual mode (for all factories)

File: api/utils/validation_utils.py [hunk 13]
```diff
@@ -723,3 +725,201 @@ def validate_id(cls, v: Any) -> str:
 
 
 class ListDatasetReq(BaseListReq): ...
+
+
+class AddModelReq(Base):
+    """Validation model for adding models (factory-level or individual model)."""
+
+    llm_factory: Annotated[str, StringConstraints(strip_whitespace=True, min_length=1), Field(...)]
+    api_key: Annotated[str | dict[str, Any] | None, Field(default=None)]
```
Comment (as-ondewo) [inline api/utils/validation_utils.py:734]: Why can this be a dict?

File: api/utils/validation_utils.py [hunk 14]
```diff
@@ -723,3 +725,201 @@ def validate_id(cls, v: Any) -> str:
 
 
 class ListDatasetReq(BaseListReq): ...
+
+
+class AddModelReq(Base):
```
Comment (as-ondewo) [inline api/utils/validation_utils.py:730]: strip whitespace for all string fields

File: test/testcases/test_http_api/test_user_models_management/test_add_model.py [hunk 15]
```diff
@@ -0,0 +1,756 @@
...
+
+    @pytest.mark.p1
+    def test_individual_model_missing_model_type(self, HttpApiAuth: RAGFlowHttpApiAuth) -> None:
+        """Test that model_type is required when llm_name is provided"""
+        res: Dict[str, Any] = add_model(HttpApiAuth, {"llm_factory": "Ollama", "llm_name": "llama2"})
+        assert res["code"] == RetCode.ARGUMENT_ERROR, res
+        assert "model_type is required when adding an individual model" in res["message"], res
+
+    @pytest.mark.p1
+    def test_individual_model_invalid_model_type(self, HttpApiAuth: RAGFlowHttpApiAuth) -> None:
```
Comment (as-ondewo) [inline test/testcases/test_http_api/test_user_models_management/test_add_model.py:86]: also test the success case for all of valid model types that should work

File: test/testcases/test_http_api/test_user_models_management/test_add_model.py [hunk 16]
```diff
@@ -0,0 +1,756 @@
...
+        ids=["empty_auth", "invalid_api_token"],
+    )
+    def test_invalid_auth(self, invalid_auth: Optional[RAGFlowHttpApiAuth], expected_code: int, expected_message: str) -> None:
+        res: Dict[str, Any] = add_model(invalid_auth, {"llm_factory": "OpenAI", "api_key": "test-key"})
+        assert res["code"] == expected_code, res
+        assert res["message"] == expected_message, res
+
+
+@pytest.mark.usefixtures("cleanup_added_models")
+class TestAddModelParameterValidation:
```
Comment (as-ondewo) [inline test/testcases/test_http_api/test_user_models_management/test_add_model.py:47]: also test:
* `llm_name` maximum length
* `api_key` required for adding all models
* `base_url` required for model without default base URL
* `max_tokens` range of valid values
